### ISSUE TITLE
Browser console warnings with React and Vite

### ISSUE THREAD

**Issue by bedwards-cms** [apostrophecms/sanitize-html#639]
## The problem to solve

Lots of console warnings are being logged since adding sanitize-html to a React application running with vite.

This is details about the warning we are seeing.
https://vitejs.dev/guide/troubleshooting.html#module-externalized-for-browser-compatibility

Here is an example what we see in console:
![image](https://github.com/apostrophecms/sanitize-html/assets/135224593/697eaf85-8a95-4f3e-a80b-d73bf46c1339)

## Proposed solution

See if there is a reasonable resolution to eliminate this issue.

## Alternatives

In this case, one alternative is to just switch to using domPurify

## Additional context

> Mention any other context or attach screenshots regarding the feature request.
> Check the existing issue labels and add any you think are relevant, e.g.,
> `UI`, `UX`, `accessibility`, `security`.

**Comment by boutell** [comment]
We don't use this module browser-side. I'm aware of some valid use cases, but it mostly makes much more sense to sanitize things in a place you can trust (e.g. on a server). So this is an area for community contributions.

**Comment by boutell** [comment]
Possibly it's time for a new major version that introduces a backwards compatibility break by making postcss support optional via a second, optional module. That seems to be source of just about all concerns about this. There is no mention of the `fs` module in sanitize-html itself.

**Comment by devCodeHub-star** [comment]
Hello @boutell 

I am also getting all the warnings in my console. I have update the version from '2.7.3' to '2.12.1' but still I am getting all warnings.
![image](https://github.com/apostrophecms/sanitize-html/assets/148462347/2a356e0b-137a-4f70-af0b-b176c381f983)
![image](https://github.com/apostrophecms/sanitize-html/assets/148462347/f3b18dce-5837-4e33-b017-e32bd1de8a80)

Any solution ?

Thankyou

**Comment by KyleBrown-804** [comment]
This might be a dumb question but an important to ask nonetheless, if I am seeing these warnings (on a Vue/Nuxt 3 application), is there any concern about the functionality of the sanitizing working properly? As a followup and per @boutell's comment, would it be safer and better practice to move the sanitizing to an SSR context?

**Comment by boutell** [comment]
Really you have two options: you can sanitize the data once server side and save it in sanitized form, or you can sanitize it every time you output it on the client side. Here we prefer the former, because it is efficient and it's harder to do something unsafe, and so we don't use this module on the client side. Client side support is contributed by the community.

These warnings relate to Node.js features utilized by sanitize-html or its dependencies. Vite is telling you that it provided a simulation, but not a complete one, and there's a chance the module uses something it doesn't provide. Possibly that could be fixed by community contributors through eliminating calls to the mentioned features.

For what it's worth though, `sanitize-html` never does import `path`. It's not mentioned anywhere in the source. So it's probably a dependency of a sub-dependency, and that's a longer road tracking down the module that actually has the issue and contributing there or finding an alternative to that functionality for sanitize-html to use.

**Comment by KyleBrown-804** [comment]
Awesome, thank you for the insight 👍

**Comment by blittle** [comment]
Related, but sanitize-html has problems running on cloudflare workers, which is also a server environment, but does not have Node APIs.

**Comment by boutell** [comment]
It sounds like this would need to be resolved upstream in `postcss`, so I would recommend checking into whether I'm right about that, and if so contributing to the solution there.

**Comment by vhscom** [comment]
Linking to https://github.com/vitejs/vite/issues/7821

**Comment by vhscom** [comment]
Switching to DOMPurify as OP suggested resolves the wall of warnings.

**Comment by vhscom** [comment]
Switching to DOMPurify as OP suggested resolved the wall of warnings with Vite 5. DOMPurify is DOM-only. If you're looking for an isomorphic solution in a single package, [js-xss](https://github.com/leizongmin/js-xss) will also work when pre-rendering on the server.

**Comment by phil294** [comment]
You can use a polyfill:

https://github.com/davidmyersdev/vite-plugin-node-polyfills

```js
// vite.config.js
plugins: [
nodePolyfills({
            include: ['process', 'tty', 'path', 'fs'],
            protocolImports: false,
        }),
```
Amounts to 69 KB bundle size increase / 16 KB gzipped

**Comment by jhoermann** [comment]
> You can use a polyfill:
> 
> https://github.com/davidmyersdev/vite-plugin-node-polyfills
> 
> // vite.config.js
> plugins: [
> nodePolyfills({
>             include: ['process', 'tty', 'path', 'fs'],
>             protocolImports: false,
>         }),
> 
> Amounts to 69 KB bundle size increase / 16 KB gzipped

That's the solution I found, too. But I'm still getting:
```
Module "source-map-js" has been externalized for browser compatibility. Cannot access "source-map-js.SourceMapConsumer" in client code. See https://vite.dev/guide/troubleshooting.html#module-externalized-for-browser-compatibility for more details.
```

**Comment by viveleroi** [comment]
It looks to me like the vite polyfill plugin works for everything in the warning messages except `source-map-js` which I believe is due to sanitize-html importing `postcss` https://github.com/apostrophecms/sanitize-html/blob/main/index.js#L6

Postcss requires source-map-js.

**Comment by jhoermann** [comment]
Still an issue.

**Comment by boutell** [comment]
We don't support browser-side use in-house. Anyone interested in contributing to a solution should look at the "browser" section of the postcss docs:

https://github.com/postcss/postcss?tab=readme-ov-file#browser

Perhaps postcss-js can be used to solve the problem, perhaps not.

**Comment by boutell** [comment]
Another possibility is a new major version in which CSS support is not present by default and must be added via a separate module. That is a refactor we'd also consider a PR for.

**Comment by boutell** [comment]
As mentioned, we won't be addressing this in-house because we don't use this module on the browser side, but PRs are welcome. Here's what you need to know.

Anyone interested in contributing to a solution should look at the "browser" section of the postcss docs:

https://github.com/postcss/postcss?tab=readme-ov-file#browser

Perhaps postcss-js can be used to solve the problem, perhaps not.

Another possibility is a new major version in which CSS support is not present by default and must be added via a separate module. That is a refactor we'd also consider a PR for.

**Comment by tgelu** [comment]
I think the confusion comes from the docs, where it appears like this is usable in the browser also

**Comment by boutell** [comment]
Yes, it is supposed to be based on past community contributions, but so far no one has stepped up to patch this issue.

**Comment by hanool** [comment]
> Same here. Nuxt v4 + Vite. Any updates?

I migrated to [DOMPurify](https://github.com/cure53/DOMPurify) for browser use and I worked great, which I found the [recommendation on MDN documents](https://developer.mozilla.org/en-US/docs/Mozilla/Add-ons/WebExtensions/Safely_inserting_external_content_into_a_page#html_sanitization).
